util/http/retry.go

```diff
@@ -89,3 +93,50 @@ func RetryStrategy(ctx context.Context, resp *http.Response, err error) (bool, e
 	}
 	return retry, err2
 }
+
+// ShouldReauthenticate takes an HTTP response and determines whether or not
+// it warrants re-authentication.
+func ShouldReauthenticate(resp *http.Response) bool {
+	switch resp.StatusCode {
+	case http.StatusUnauthorized:
+		return true
+	case http.StatusForbidden:
+
+		// Although in most cases 403 responses represent authorization issues that generally
+		// cannot be resolved by re-authentication, some registries like ECR, will return a 403 on
+		// credential expiration. (ref https://docs.aws.amazon.com/AmazonECR/latest/userguide/common-errors-docker.html#error-403)
+
+		// In the case of ECR, the response body is structured according to the error format defined in the
+		// Docker v2 API spec. (ref https://distribution.github.io/distribution/spec/api/#errors).
+		// We will attempt to decode the response body as a `docker.Errors`. If it can be decoded,
+		// we will ensure that the `Message` represents token expiration.
+
+		// Since we drain the response body, we will copy it to a
+		// buffer and re-assign it so that callers can still read
+		// from it.
+		body, err := io.ReadAll(resp.Body)
+		defer func() {
+			resp.Body = io.NopCloser(bytes.NewReader(body))
```

Comment:
Should we close the old instance of `resp.Body` before reinstantiating with a new `ReadCloser`? I'm wondering if no doing so would result in the TCP connection not being reused.

Reply:
Yes, we should, good call.